### PR TITLE
Gh 1186 improve parquet ingest

### PR DESCRIPTION
**Pull request by ac74475**

### CODE REVIEW

File: store-implementation/parquet-store/README.md [hunk 1]
```diff
@@ -61,13 +61,16 @@ The `ParquetStoreProperties` class contains all properties relating to the confi
...
-- `parquet.temp_data.dir`: The file path to save the temporary graph files under, by default this will be a relative path \<current path\>/.gaffer/temp_parquet_data. Warning: this directory will automatically be deleted at the start and end of any `AddElements` operation;
-- `parquet.add_elements.threadsAvailable`: The number of threads to make available to the `AddElements` operations to increase the parallelism, by default this is set to 3 which will provide maximum parallelism when adding a single Gaffer group;
-- `parquet.add_elements.row_group.size`: This parameter sets the maximum row group size in bytes before compression for the Parquet files, see [Parquet documentation](https://parquet.apache.org/documentation/latest/) for more information. By default this is set to 4MB;
+- `parquet.temp_data.dir`: The file path to save the temporary graph files under, by default this will be a relative path 
+\<current path\>/.gaffer/temp_parquet_data. Warning: this directory will automatically be deleted at the start and end of any `AddElements` operation;
```
Comment (gaffer01) [inline store-implementation/parquet-store/README.md]: Don't add a new line in the middle of the sentence.

File: store-implementation/parquet-store/README.md [hunk 2]
```diff
@@ -342,6 +345,17 @@ It is also worth noting that the order of the `Object[]` created by the `getParq
 `getPOJOFromObjects` method should match the ordering of the leaf nodes in the Parquet schema, so in the example the 
 raw_bytes is first and the cardinality is second.
 
+## Optimisations
+
+With Get operations there are optimisations performed to allow for the minimal number of files to be queried based on 
+the inputs to the query.
+
+This means that you will get significant improvements if you can formulate your filters with the `Or` and `Not` filters as close to 
+the leaves of the filter tree as possible.
```
Comment (gaffer01) [inline store-implementation/parquet-store/README.md]: No new line.

File: store-implementation/parquet-store/README.md [hunk 3]
```diff
@@ -342,6 +345,17 @@ It is also worth noting that the order of the `Object[]` created by the `getParq
...
+
+You will also gain significant improvements if you run two queried and merged/deduplicate the results locally rather than using 
+an `Or` filter at the root of the filter tree.
```
Comment (gaffer01) [inline store-implementation/parquet-store/README.md]: No new line.

File: store-implementation/parquet-store/README.md [hunk 4]
```diff
@@ -342,6 +345,17 @@ It is also worth noting that the order of the `Object[]` created by the `getParq
...
 
+## Optimisations
+
+With Get operations there are optimisations performed to allow for the minimal number of files to be queried based on 
+the inputs to the query.
+
+This means that you will get significant improvements if you can formulate your filters with the `Or` and `Not` filters as close to 
+the leaves of the filter tree as possible.
+
+You will also gain significant improvements if you run two queried and merged/deduplicate the results locally rather than using
```
Comment (gaffer01) [inline store-implementation/parquet-store/README.md]: Can you give an example?

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java [hunk 5]
```diff
@@ -54,26 +58,63 @@ public AggregateAndSortTempData(final ParquetStore store, final SparkSession spa
...
+                final Set<String> currentGraphFiles = new HashSet<>();
+                if (currentDataInThisGroupDir != null) {
+//                    if (splitPoints != null) {
+                        currentGraphFiles.add(currentDataInThisGroupDir + "/part-" + zeroPad(String.valueOf(i), 5) + "-*.parquet");
+//                        int x = i - 1;
+//                        while (x >= 0 && splitPoints.get(x) == null) {
+//                            currentGraphFiles.add(currentDataInThisGroupDir + "/part-" + zeroPad(String.valueOf(i), 5) + "-*.parquet");
+//                            x--;
+//                        }
+//                    }
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java]: Remove commented out lines.

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java [hunk 6]
```diff
@@ -102,4 +155,12 @@ public AggregateAndSortTempData(final ParquetStore store, final SparkSession spa
         }
         pool.shutdown();
     }
+
+    private String zeroPad(final String input, final int length) {
+        StringBuilder temp = new StringBuilder(input);
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java]: Add final

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateGroup.java [hunk 7]
```diff
@@ -134,41 +139,11 @@ public OperationException call() {
...
-                        .sort(firstSortColumn, groupBySeq.result());
 
-                // Write out aggregated and sorted data
-                sortedData
-                        .coalesce(filesPerGroup)
+                // Write out aggregated data
+                aggregatedData
                         .write()
-                        .option("compression", "gzip")
+                        .option("compression", "none")
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateGroup.java]: Maybe use the default compression?

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/CalculateSplitPointsFromIterable.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                } else {
+                    counter++;
+                }
+            }
+        }
+        if (sample.isEmpty()) {
+            return new TreeMap<>(COMPARATOR);
+        } else {
+            sample.sort(COMPARATOR);
+            final int x = (sample.size() / (numOfSplits + 1)) + 1;
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/CalculateSplitPointsFromIterable.java]: Can you use a more meaningful variable name than `x`?

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/GenerateIndexedColumnSortedData.java [hunk 9]
```diff
@@ -39,7 +39,7 @@
  */
 public class GenerateIndexedColumnSortedData implements Callable<OperationException>, Serializable {
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(uk.gov.gchq.gaffer.parquetstore.operation.addelements.impl.AggregateAndSortGroup.class);
+    private static final Logger LOGGER = LoggerFactory.getLogger(AggregateGroup.class);
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/GenerateIndexedColumnSortedData.java]: Shouldn't this be `GenerateIndexedColumnSortedData.class`?

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/GenerateIndices.java [hunk 10]
```diff
@@ -50,9 +51,14 @@
...
-        final ExecutorService pool = Executors.newFixedThreadPool(parquetStoreProperties.getThreadsAvailable());
-        final String tempFileDir = store.getTempFilesDir();
+        final int numberOfThreads;
+        final Option<String> sparkDriverCores = spark.conf().getOption("spark.driver.cores");
+        if (sparkDriverCores.nonEmpty()) {
+            numberOfThreads = Integer.parseInt(sparkDriverCores.get());
+        } else {
+            numberOfThreads = store.getProperties().getThreadsAvailable();
+        }
+        final ExecutorService pool = Executors.newFixedThreadPool(numberOfThreads);        final String tempFileDir = store.getTempFilesDir();
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/GenerateIndices.java]: `final String tempFileDir` - should be on a separate line.

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/RDD/CalculateSplitPointsFromJavaRDD.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import uk.gov.gchq.gaffer.data.element.comparison.ComparableOrToStringComparator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * Generates the split points from an {@link JavaRDD} of {@link Element}s by selecting a sample of the data, sorting that sample and
+ * then pulling out the relevant objects to act as the split points */
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/RDD/CalculateSplitPointsFromJavaRDD.java]: The `*/` should be on its own line.

File: store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/utils/SortDataTest.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+import java.io.IOException;
+
+public class SortDataTest {
+
+    @BeforeClass
+    public static void GeneratePreAggregatedData() throws IOException {
+        final SchemaUtils schemaUtils = new SchemaUtils(TestUtils.gafferSchema("schemaUsingLongVertexType"));
+        ParquetWriter<Element> writer = new ParquetElementWriter
+                .Builder(new Path(".SortDataTest/SortDataTest/graph/GROUP=" + TestGroups.ENTITY + "/aggregated/split0/part-0.parquet"))
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/utils/SortDataTest.java]: Can you use a temp folder as other tests that need to write to file do?

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/RDD/WriteUnsortedDataFunction.java [hunk 13]
```diff
@@ -37,17 +38,21 @@
     private static final long serialVersionUID = 1420859039414174311L;
     private final String tempFilesDir;
     private final byte[] gafferSchema;
+    private final Map<String, Map<Integer, Object>> groupToSplitPoints;
 
 
-    public WriteUnsortedDataFunction(final String tempFilesDir, final SchemaUtils schemaUtils) {
+    public WriteUnsortedDataFunction(final String tempFilesDir, final SchemaUtils schemaUtils,
+                                     final Map<String, Map<Integer, Object>> groupToSplitPoints,
+                                     final int numOfFilesPerGroup) {
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/RDD/WriteUnsortedDataFunction.java]: This parameter (`numOfFilesPerGroup`) isn't used anywhere.

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java [hunk 14]
```diff
@@ -54,26 +58,63 @@ public AggregateAndSortTempData(final ParquetStore store, final SparkSession spa
         } else {
             currentDataDir = null;
         }
+        final int numberOfSplits = store.getProperties().getAddElementsOutputFilesPerGroup();
         for (final String group : schemaUtils.getEdgeGroups()) {
+            final Map<Integer, Object> splitPoints = groupToSplitPoints.get(group);
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java]: This variable isn't used.

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java [hunk 15]
```diff
@@ -54,26 +58,63 @@ public AggregateAndSortTempData(final ParquetStore store, final SparkSession spa
...
+//                            currentGraphFiles.add(currentDataInThisGroupDir + "/part-" + zeroPad(String.valueOf(i), 5) + "-*.parquet");
+//                            x--;
+//                        }
+//                    }
+                }
+                tasks.add(new AggregateGroup(group, ParquetStoreConstants.SOURCE, store, currentGraphFiles, spark, i));
+            }
         }
         for (final String group : schemaUtils.getEntityGroups()) {
+            final Map<Integer, Object> splitPoints = groupToSplitPoints.get(group);
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java]: This variable isn't used.

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/WriteUnsortedData.java [hunk 16]
```diff
@@ -87,15 +99,40 @@ private void _writeElements(final Iterator<? extends Element> elements) throws O
...
-        if (fileNumber == null) {
-            groupToFileNumber.put(group, 0);
-            fileNumber = 0;
+    private ParquetWriter<Element> getWriter(final Map<Integer, ParquetWriter<Element>> splitToWriter,
+                                             final Map<Integer, Object> splitPoints,
+                                             final Object gafferObject, final String group, final String column) throws IOException {
+        final int numOfSplits = splitPoints.size();
+        for (int i = 1; i < numOfSplits; i++) {
+            final Object splitPoint = splitPoints.get(i);
+            final int comparision = COMPARATOR.compare(gafferObject, splitPoint);
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/WriteUnsortedData.java]: `comparision` should be `comparison`.

**Comment by ac74475** [comment]
Closing while changing the way the sorting is done, so it just sorts within each split partition

File: store-implementation/parquet-store/README.md [hunk 17]
```diff
@@ -62,17 +63,15 @@ The `ParquetStoreProperties` class contains all properties relating to the confi
...
-for more information. By default this is set to 1MB;
-- `parquet.add_elements.output_files_per_group`: This is the number of files that the output data is split into per Gaffer group. By default this is set to 100.
+- `parquet.add_elements.page.size`: This just exposes the Parquet file format parameter controlling the maximum page and dictionary page size in bytes before compression, see [Parquet documentation](https://parquet.apache.org/documentation/latest/) for more information. By default this is set to 1MB;
+- `parquet.add_elements.output_files_per_group`: This is the number of files that the output data is split into per Gaffer group. By default this is set to 10.
 
 A complete Gaffer properties file using a `ParquetStore` will look like:
 
 ```
-gaffer.store.class=uk.gov.gchq.gaffer.parquetstore.ParquetStore
+gaffer.store.class=uk.gov.gchq.gaffer.parquetstore.ParquetStore````
```
Comment (gaffer01) [inline store-implementation/parquet-store/README.md]: Those 4 ` at the end of the line should be deleted.

File: store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/utils/WriteUnsortedDataTest.java [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        final Map<String, Map<Object, Integer>> splitPoints = new HashMap<>(2);
+        splitPoints.put(TestGroups.ENTITY, new CalculateSplitPointsFromIterable(2, 2).calculateSplitsForGroup(data, TestGroups.ENTITY, true));
+        splitPoints.put(TestGroups.ENTITY_2, new CalculateSplitPointsFromIterable(2, 2).calculateSplitsForGroup(data, TestGroups.ENTITY_2, true));
+        final WriteUnsortedData writeUnsortedData = new WriteUnsortedData(props.getTempFilesDir() + "/WriteUnsortedDataTest", schemaUtils, splitPoints);
+        writeUnsortedData.writeElements(data.iterator());
+        final String Entity1Split0 = props.getTempFilesDir() + "/WriteUnsortedDataTest/graph/GROUP=" + TestGroups.ENTITY + "/raw/split0";
+        final String Entity1Split1 = props.getTempFilesDir() + "/WriteUnsortedDataTest/graph/GROUP=" + TestGroups.ENTITY + "/raw/split1";
+        final String Entity2Split0 = props.getTempFilesDir() + "/WriteUnsortedDataTest/graph/GROUP=" + TestGroups.ENTITY_2 + "/raw/split0";
+        final String Entity2Split1 = props.getTempFilesDir() + "/WriteUnsortedDataTest/graph/GROUP=" + TestGroups.ENTITY_2 + "/raw/split1";
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/utils/WriteUnsortedDataTest.java]: The above 4 variables should start with a lower-case 'e'.

File: store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/utils/WriteUnsortedDataTest.java [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        final Map<String, Map<Object, Integer>> splitPoints = new HashMap<>(2);
+        splitPoints.put(TestGroups.EDGE, new CalculateSplitPointsFromIterable(2, 2).calculateSplitsForGroup(data, TestGroups.EDGE, false));
+        splitPoints.put(TestGroups.EDGE_2, new CalculateSplitPointsFromIterable(2, 2).calculateSplitsForGroup(data, TestGroups.EDGE_2, false));
+        final WriteUnsortedData writeUnsortedData = new WriteUnsortedData(props.getTempFilesDir() + "/WriteUnsortedDataTest", schemaUtils, splitPoints);
+        writeUnsortedData.writeElements(data.iterator());
+        final String Edge1Split0 = props.getTempFilesDir() + "/WriteUnsortedDataTest/graph/GROUP=" + TestGroups.EDGE + "/raw/split0";
+        final String Edge1Split1 = props.getTempFilesDir() + "/WriteUnsortedDataTest/graph/GROUP=" + TestGroups.EDGE + "/raw/split1";
+        final String Edge2Split0 = props.getTempFilesDir() + "/WriteUnsortedDataTest/graph/GROUP=" + TestGroups.EDGE_2 + "/raw/split0";
+        final String Edge2Split1 = props.getTempFilesDir() + "/WriteUnsortedDataTest/graph/GROUP=" + TestGroups.EDGE_2 + "/raw/split1";
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/utils/WriteUnsortedDataTest.java]: The above 4 variables should start with a lower-case 'e'.

File: store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/utils/SortDataTest.java [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+import uk.gov.gchq.gaffer.parquetstore.testutils.DataGen;
+import uk.gov.gchq.gaffer.parquetstore.testutils.TestUtils;
+import uk.gov.gchq.gaffer.store.StoreException;
+
+import java.io.IOException;
+
+public class SortDataTest {
+
+    @BeforeClass
+    public static void GeneratePreAggregatedData() throws IOException {
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/utils/SortDataTest.java]: Lower-case 'g'

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java [hunk 21]
```diff
@@ -56,25 +68,50 @@ public AggregateAndSortTempData(final ParquetStore store, final SparkSession spa
...
-            tasks.add(new AggregateAndSortGroup(group, ParquetStoreConstants.SOURCE, store, currentDataInThisGroupDir, spark));
         }
         for (final String group : schemaUtils.getEntityGroups()) {
-            final String currentDataInThisGroupDir;
-            if (currentDataDir != null) {
-                currentDataInThisGroupDir = ParquetStore.getGroupDirectory(group, ParquetStoreConstants.SOURCE, currentDataDir);
-            } else {
-                currentDataInThisGroupDir = null;
+            if (groupToSplitPoints.containsKey(group)) {
+                final String currentDataInThisGroupDir;
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/impl/AggregateAndSortTempData.java]: The code inside this 'if' statement is almost identical to the code in the 'if' statement above - can you refactor it into a separate method and re-use?

**Pull request merged**